### PR TITLE
core/filtermaps: fix map renderer reorg issue

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR fixes a bug in the map renderer that sometimes used an obsolete block log value pointer to initialize the iterator for rendering from a snapshot. This bug was triggered by chain reorgs and sometimes caused indexing errors and invalid search results. A few other conditions are also made safer that were not reported to cause issues yet but could potentially be unsafe in some corner cases. A new unit test is also added that reproduced the bug but passes with the new fixes.

Fixes https://github.com/ethereum/go-ethereum/issues/31593

### CODE REVIEW

File: core/filtermaps/map_renderer.go [hunk 1]
```diff
@@ -257,10 +256,13 @@ func (f *FilterMaps) loadHeadSnapshot() error {
...
-		filterMap:     r.currentMap.filterMap.copy(),
+	if r.iterator.blockNumber != r.currentMap.lastBlock {
+		panic("iterator state inconsistent with last block")
+	}
+	r.f.renderSnapshots.Add(r.currentMap.lastBlock, &renderedMap{
+		filterMap:     r.currentMap.filterMap.fastCopy(),
 		mapIndex:      r.currentMap.mapIndex,
-		lastBlock:     r.iterator.blockNumber,
+		lastBlock:     r.currentMap.lastBlock,
 		lastBlockId:   r.f.targetView.getBlockId(r.currentMap.lastBlock),
```
Comment (rjl493456442) [inline core/filtermaps/map_renderer.go:266]: Honestly it's a bit weird to get the block id with `r.f.targetView`, `r.iterator.chainView` might be a better candidate.

**Comment by rjl493456442** [comment]
Generally lgtm. Could you please point out which part of code was wrong before?

File: core/filtermaps/filtermaps.go [hunk 2]
```diff
@@ -138,16 +139,31 @@ type FilterMaps struct {
...
 type filterMap []FilterRow
 
-// copy returns a copy of the given filter map. Note that the row slices are
-// copied but their contents are not. This permits extending the rows further
+// fastCopy returns a copy of the given filter map. Note that the row slices are
+// copied but their contents are not. This permits appending to the rows further
 // (which happens during map rendering) without affecting the validity of
 // copies made for snapshots during rendering.
-func (fm filterMap) copy() filterMap {
+// Appending to the rows of both the original map and the fast copy, or two fast
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply core/filtermaps/filtermaps.go:146]: Technically true but the point is you should not append to both as they share the underlying array and would corrupt each other. In my use case the original keeps being appended as it is the current rendered map and snapsnots are made during the process. So I defined the semantics of `fastCopy` so that the original can be appended to and the copy should be used as read only. This rule guarantees safety.

**Comment by zsfelfoldi** [comment]
> Generally lgtm. Could you please point out which part of code was wrong before?

It was `newLogIteratorFromBlockDelimiter` which should initialize an iterator at a block boundary. It is used when there is a recent snapshot available. Practically it is used every time we are adding a new block to the head of the existing chain, or if there was a small reorg but a valid snapshot is still available. It needs the log value pointer to the block boundary _after_ the snapshot block so that it can resume rendering from there. Now this is available in the snapshot (`renderedMap.headDelimiter`) but eariler it wasn't stored there and originally I implemented it in `newLogIteratorFromBlockDelimiter` so that it only received the block number and determined the pointer after the block itself. In the head append case this came from `filterMapsRange.headDelimiter` but in the reorg case it did read `getBlockLvPointer(blockNumber + 1)`. Now if it's reorging to previously unseen fork then it's fine because then the snapshot is a common ancestor and the pointer after it is correct in the database. But if it's reorging back to a previously processed fork (like `A->B`, then `A->B2`, then `A->B2->C2`, then `A->B->C`) then it will use the snapshot of `B` but initialize the pointer before `C2` which will be incorrect if `B` and `B2` generated a different number of log values.
Now it's simpler because `headDelimiter` is already available with every checkpoint and always consistent with the right fork, I just forgot to change `newLogIteratorFromBlockDelimiter` so that it actually uses that.

File: core/filtermaps/map_renderer.go (hunk 1 above)
Comment (zsfelfoldi) [reply core/filtermaps/map_renderer.go]: You are right, it does look better. It is an assumption that the two should always be the same, guaranteed by `iterator.updateChainView` that is always called after `r.f.targetView` could have been changed. But it is better to explicitly check rather that quietly assume so I also added a check.
